src/app/clusters/tls-certificate-management-server/CertificateTableImpl.cpp

```diff
@@ -38,7 +41,294 @@ namespace {
+        ReturnErrorOnFailure(
+            GetNextCertificateId(mNextClientId, kMaxClientCertId, mClientCertMappingCount, mClientCertMapping.data()));
+        return ConsumeId(fabric, id, mNextClientId, mClientCertMappingCount, mClientCertMapping.data());
+    }
+
+    CHIP_ERROR GetNextRootCertificateId(FabricIndex fabric, TLSCAID & id)
+    {
+        VerifyOrReturnError(mRootCertMappingCount < mRootCertMapping.size(), CHIP_ERROR_NO_MEMORY);
+        ReturnErrorOnFailure(GetNextCertificateId(mNextRootId, kMaxRootCertId, mRootCertMappingCount, mRootCertMapping.data()));
+        return ConsumeId(fabric, id, mNextRootId, mRootCertMappingCount, mRootCertMapping.data());
+    }
+
+    CHIP_ERROR RemoveClientCertificate(PersistentStorageDelegate & storage, ClientCertificateTable & table, FabricIndex fabric,
+                                       TLSCCDID id)
+    {
+        return DoRemoval(storage, table, fabric, id, mClientCertMappingCount, mClientCertMapping.data());
+    }
+
+    CHIP_ERROR RemoveRootCertificate(PersistentStorageDelegate & storage, RootCertificateTable & table, FabricIndex fabric,
+                                     TLSCAID id)
+    {
+        return DoRemoval(storage, table, fabric, id, mRootCertMappingCount, mRootCertMapping.data());
+    }
+
+private:
+    template <class CertificateTable>
+    CHIP_ERROR DoRemoval(PersistentStorageDelegate & storage, CertificateTable & table, FabricIndex fabric, uint16_t id,
+                         size_t count, StoredCertificate * source)
+    {
+        // Find the entry in the global mapping
+        StoredCertificate * foundEntry = nullptr;
+        for (size_t i = 0; i < count; ++i)
+        {
+            auto & entry = source[i];
+            if (entry.fabric == fabric && entry.id == id)
+            {
+                foundEntry = &entry;
+                break;
+            }
+        }
+        if (foundEntry == nullptr)
+        {
+            return CHIP_ERROR_NOT_FOUND;
+        }
+        foundEntry->fabric = kUndefinedFabricIndex;
+        foundEntry->id.Clear();
+        ReturnErrorOnFailure(this->Save(&storage));
+
+        CertificateId localId(id);
+        auto removeResult = table.RemoveTableEntry(fabric, localId);
+        if (removeResult == CHIP_ERROR_NOT_FOUND)
+        {
+            // Index is out of sync; keep it updated & return NOT_FOUND
+            return CHIP_ERROR_NOT_FOUND;
+        }
+
+        if (removeResult != CHIP_NO_ERROR)
+        {
+            // Failed to remove, re-add
+            foundEntry->fabric = fabric;
+            foundEntry->id     = localId;
+            ReturnErrorOnFailure(this->Save(&storage));
+        }
+        return removeResult;
+    }
+
+    CHIP_ERROR GetNextCertificateId(uint16_t & nextId, uint16_t maxId, size_t count, const StoredCertificate * source)
+    {
+        bool looped = false;
+        bool taken;
+        uint16_t outOfIdCheck = nextId;
+        do
+        {
+            if (nextId == outOfIdCheck)
```

Comment:
This will always test true, so this is still broken....

I was serious about reverting to the previous, working, version and trying to do anything involving `outOfIdCheck` in a followup.

Reply:
Ah, I see.  Yes, agreed it's correct, but somewhat confusing to follow...